sdk/identity/Azure.Identity/src/AzureApplicationCredential.cs

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core;
+using Azure.Core.Pipeline;
+
+namespace Azure.Identity
+{
+    /// <summary>
+    /// Provides a <see cref="TokenCredential"/> implementation which chains the <see cref="EnvironmentCredential"/> and <see cref="ManagedIdentityCredential"/> implementations to be tried in order
+    /// until one of the getToken methods returns a non-default <see cref="AccessToken"/>.
+    /// </summary>
+    /// <remarks>
+    /// This credential is designed for applications deployed to Azure <see cref="DefaultAzureCredential"/> is
+    /// better suited to local development). It authenticates service principals and managed identities..
+    /// </remarks>
+    public class AzureApplicationCredential : TokenCredential
+    {
+        private readonly ChainedTokenCredential _credential;
+
+        /// <summary>
+        /// Constructor for instrumenting in tests
+        /// </summary>
+        protected AzureApplicationCredential()
+        { }
+
+        /// <summary>
+        /// Initializes an instance of the <see cref="AzureApplicationCredential"/>.
+        /// </summary>
+        /// <param name="options">The <see cref="TokenCredentialOptions"/> to configure this credential.</param>
+        public AzureApplicationCredential(AzureApplicationCredentialOptions options) : this(options, null, null)
+        { }
+
+        internal AzureApplicationCredential(AzureApplicationCredentialOptions options, EnvironmentCredential environmentCredential = null, ManagedIdentityCredential managedIdentityCredential = null)
+        {
+            Argument.AssertNotNull(options, nameof(options));
+            Argument.AssertNotNull(options.ManagedIdentityClientId, $"{nameof(AzureApplicationCredentialOptions)}.{nameof(options.ManagedIdentityClientId)}");
+
+            _credential = new ChainedTokenCredential(
+                environmentCredential ?? new EnvironmentCredential(options),
+                managedIdentityCredential ?? new ManagedIdentityCredential(options.ManagedIdentityClientId)
```

Comment:
If `ManagedIdentityClientId` is null we should call the `ManagedIdentityCredential` ctor overload which takes no arguments.